File: java/src/main/java/generating/factory/MainFactory.java
package generating.factory;

enum CarType {
	PassengerCar("Passenger car"), SUV("SUV");

	private String type;

	CarType(String type) {
		this.type = type;
	}

	public String getDescription() {
		return this.type;
	}
}

interface Car {
	public String getCarType();
}

class PassengerCar implements Car {
	@Override
	public String getCarType() {
		return CarType.PassengerCar.getDescription();
	}
}

class SUV implements Car {
	@Override
	public String getCarType() {
		return CarType.SUV.getDescription();
	}
}

class CarFactory {
	Car getCar(final CarType type) {
		switch (type) {
		case PassengerCar:
			return new PassengerCar();
		case SUV:
			return new SUV();
		default:
			return null;
		}
	}
}

public class MainFactory {
	public static void main(String[] args) {
		CarFactory carFactory = new CarFactory();
		Car car = carFactory.getCar(CarType.PassengerCar);
		System.out.println(car.getCarType());
		car = carFactory.getCar(CarType.SUV );
		System.out.println(car.getCarType());
	}
}

File: kotlin/src/main/kotlin/generating/factory/MainFactory.kt
package generating.factory

enum class CarType {
    PassengerCar("Passenger car"), SUV("SUV");

    private val type: String;

    constructor(type: String) {
        this.type = type;
    }

    public fun getDescription(): String {
        return this.type;
    }
}

internal interface Car {
    val carType: String
}

internal class PassengerCar : Car {
    override val carType: String
        get() = CarType.PassengerCar.getDescription()
}

internal class SUV : Car {
    override val carType: String
        get() = CarType.SUV.getDescription()
}

internal class CarFactory {
    fun getCar(type: CarType): Car? {
        return when (type) {
            CarType.PassengerCar -> PassengerCar()
            CarType.SUV -> SUV()
            else -> null
        }
    }
}

fun main() {
    val carFactory = CarFactory()
    var car = carFactory.getCar(CarType.PassengerCar)
    println(car?.carType)
    car = carFactory.getCar(CarType.SUV)
    println(car?.carType)
}